Rewrite docker conceptual tutorial (#11756)
dotnet/docs: docs/core/docker/build-container.md

---
title: Containerize an app with Docker tutorial
description: In this tutorial, you will learn how to containerize a .NET Core application with Docker.

> ms.date: 03/20/2019

Reviewer: Do you want to update the date?

ms.topic: tutorial
ms.custom: "mvc, seodec18"
#Customer intent: As a developer, I want to containerize my .NET Core app so that I can deploy it to the cloud.
---

# Tutorial: Containerize a .NET Core app

This tutorial teaches you how to build a Docker image that contains your .NET Core application. The image can be used to create containers for your local development environment, private cloud, or public cloud.

> In this tutorial you'll learn how to:

Reviewer: My rant: I find starting the first two paragraphs with "this tutorial" repetitive and boilerplate. Consider:

```suggestion
You'll learn to:
```

> [!div class="checklist"]
> * Create and publish a simple .NET Core app
> * Create and configure a Dockerfile for .NET Core
> * Build a Docker image
> * Create and run a Docker container

> This tutorial teaches the Docker container build and deploy tasks for a .NET Core application. The *Docker platform* uses the *Docker engine* to quickly build and package apps as *Docker images*. These images are written in the *Dockerfile* format to be deployed and run in a layered container.

Reviewer: Again with "this tutorial" (and I know it's in the template. It's still repetitive.)

```suggestion
You'll understand the Docker container build and deploy tasks for a .NET Core application. The *Docker platform* uses the *Docker engine* to quickly build and package apps as *Docker images*. These images are written in the *Dockerfile* format to be deployed and run in a layered container.
```

## Prerequisites

> To complete this tutorial, install the following prerequisites:

Reviewer: ```suggestion
Install the following prerequisites:
```

## Create .NET Core app

You need a .NET Core app that the Docker container will run. Open your terminal, create a working directory, and enter it. In the working directory, run the following command to create a new project in a subdirectory named app:

```console
dotnet new console -o app -n myapp
```

That command creates a new directory named *app* and generates a "Hello World" app. You can test this app to see what it does. Enter the *app* directory and execute the command `dotnet run`. You'll see the following output:

```console
> dotnet run
Hello World!
```

The default template creates an app that prints to the terminal and then exits. For this tutorial, you'll use an app that loops indefinitely. Open the **Program.cs** file in a text editor. It should currently look like the following code:

```csharp
using System;

namespace myapp
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}
```

Replace the file with the following code that counts numbers every second:

```csharp
using System;

namespace myapp
{
    class Program
    {

>         static void Main(string[] args)

Reviewer: This feature is from 7.1, so should be available for any user.

```suggestion
        static async Task Main(string[] args)
```

Author: the default template for 2.2 seems to target 7.0 as the language :(

Reviewer: Yes, but you can set it with this in your csproj:

```XML
<LangVersion>latest</LangVersion>
```

And, that will be changing with the next .NET Core SDK (where latest or preview is the default).  See here: https://docs.microsoft.com/dotnet/csharp/language-reference/configure-language-version

Author: Is there a `dotnet` command to do this? I would rather not have to go into more steps about the csproj file. That just adds more stuff for the user to read about the sample app when the intent of the article is about getting the app into the docker container.

Reviewer: Not that I know about. You can ignore this suggestion, if you prefer. I agree it's not the focus.

        {
            var counter = 0;
            var max = args.Length != 0 ? Convert.ToInt32(args[0]) : -1;
            while(max == -1 || counter < max)
            {
                counter++;
                Console.WriteLine($"Counter: {counter}");

>                 System.Threading.Tasks.Task.Delay(1000).Wait();

Reviewer: ```suggestion
                await System.Threading.Tasks.Task.Delay(1000);
```

            }
        }
    }
}
```

Save the file and test the program again with `dotnet run`. Remember that this app runs indefinitely. Use the cancel command <kbd>CTRL + C</kbd> to stop it. You'll see the following output:

```console
> dotnet run
Counter: 1
Counter: 2
Counter: 3
Counter: 4
^C
```

If you pass a number on the command line to the app, it will only count up to that amount and then exit. Try it with `dotnet run -- 5` to count to five.

> [!NOTE]
> Any parameters after `--` are passed to your application.

> ### Using a newer SDK

Reviewer: Suggestion: move this section above, right after pre-requisites. The `dotnet --info` will tell the reader what SDK they are using, and if a `global.json` is determining the current SDK selection. It might flow better.

```suggestion
### Use SDK version 2.2
```

If you're using an SDK that is newer, like a 3.0 preview, make sure that your app is forced to use the 2.2 SDK. Create a file named `global.json` in your working directory and paste in the following json code:

```json
{                                             
  "sdk": {                                    
    "version": "2.2.100"                      
  }                                           
}                                             
```

Save this file. The presence of file will force .NET Core to use version 2.2 for any `dotnet` command called from this directory and below.

## Publish .NET Core app

> Before you add your .NET Core app to the Docker image, publish it. The published version of the app is what the container will execute when it's started.

Reviewer: ```suggestion
Before you add your .NET Core app to the Docker image, publish it. The container will execute the published version of the app when it's started.
```

From the working directory, enter the **app** directory with the example source code and run the following command:

```console
dotnet publish -c Release
```

This command compiles your app to the **publish** folder in the output folder of your app. The path to the **publish** folder from the working directory should be `.\app\bin\Release\netcoreapp2.2\publish\`

Get a directory listing of the publish folder to verify that the **myapp.dll** was created. From the **app** directory, run one of the following commands:

```console
> dir bin\Release\netcoreapp2.2\publish
 Directory of C:\path-to-working-dir\app\bin\Release\netcoreapp2.2\publish

04/05/2019  11:00 AM    <DIR>          .
04/05/2019  11:00 AM    <DIR>          ..
04/05/2019  11:00 AM               447 myapp.deps.json
04/05/2019  11:00 AM             4,608 myapp.dll
04/05/2019  11:00 AM               448 myapp.pdb
04/05/2019  11:00 AM               154 myapp.runtimeconfig.json
```

```bash
me@DESKTOP:/path-to-working-dir/app$ ls bin/Release/netcoreapp2.2/publish
myapp.deps.json  myapp.dll  myapp.pdb  myapp.runtimeconfig.json
```

> In your terminal, go back a directory to the working directory.

Reviewer: ```suggestion
In your terminal, go up a directory to the working directory.
```

```

The `FROM` command tells Docker to pull down the image tagged **2.2** from the **mcr.microsoft.com/dotnet/core/runtime** repository. Make sure that you pull the .NET Core runtime that matches the runtime targeted by your SDK. For example, the app created in the previous section used the .NET Core 2.2 SDK and created an app that targeted .NET Core 2.2. So the base image referred to in the *Dockerfile* is tagged with **2.2**.

Save the file. From your terminal, run `docker build -t myimage .` and Docker will process each line in the *Dockerfile*. The `.` in the `docker build` command tells docker to use the current directory to find a *Dockerfile*. This command builds the image and creates a local repository named **myimage** that points to that image. After this command finishes, run `docker images` to see a list of images installed:

```console
> docker images
REPOSITORY                              TAG                 IMAGE ID            CREATED             SIZE
mcr.microsoft.com/dotnet/core/runtime   2.2                 d51bb4452469        2 days ago          314MB
myimage                                 latest              d51bb4452469        2 days ago          314MB
```

> Notice that the two images share the same **IMAGE ID** value. The value is the same between both images because the only command in the *Dockerfile* was to base the new image on an existing image. Comparing the two images, they're exactly the same. Let's add two commands to the *Dockerfile*. Each command will create a new image layer with the final command representing the image the **myimage** repository will point to.

Reviewer: The following sentence is redundant.

```suggestion
Notice that the two images share the same **IMAGE ID** value. The value is the same between both images because the only command in the *Dockerfile* was to base the new image on an existing image. Let's add two commands to the *Dockerfile*. Each command creates a new image layer with the final command representing the image the **myimage** repository will point to.
```

> Notice that the two images share the same **IMAGE ID** value. The value is the same between both images because the only command in the *Dockerfile* was to base the new image on an existing image. Comparing the two images, they're exactly the same. Let's add two commands to the *Dockerfile*. Each command will create a new image layer with the final command representing the image the **myimage** repository will point to.

Reviewer: Is **myimage** a repository, or an image?

Author: I would liken it to a git branch vs git commit ID. The git branch points to a git commit. A `repository:tag` combination points to an image ID. If you omit the tag in referencing an image, it defaults to `repository:latest`. So when you use build to create an image `-t` assigns the repository name and since `:tag` was omitted, it's given `:latest`.

```dockerfile
COPY app\bin\Release\netcoreapp2.2\publish\ .\app

ENTRYPOINT ["dotnet" "app/myapp.dll"]
```

The `COPY` command tells Docker to copy the specified folder on your computer to a folder in the container. In this example, the **publish** folder is copied to a folder named **app** in the container.

> The next command, `ENTRYPOINT`, tells docker to configure the container to run as an executable. After you start a container based on this image, it will automatically stop when the process identified from the `ENTRYPOINT` command ends.

Reviewer: You say the container stops when the program ends. I'd add a sentence before that (or modify the first sentence) to say that the container starts running the program identified by `ENTRYPOINT`.

 ---> d51bb4452469
Step 2/3 : COPY app/bin/Release/netcoreapp2.2/publish/ app/
 ---> a1e98ac62017
Step 3/3 : ENTRYPOINT ["dotnet", "app/myapp.dll"]
 ---> Running in f34da5c18e7c
Removing intermediate container f34da5c18e7c
 ---> ddcc6646461b
Successfully built ddcc6646461b
Successfully tagged myimage:latest

> docker images
REPOSITORY                              TAG                 IMAGE ID            CREATED             SIZE
myimage                                 latest              ddcc6646461b        10 seconds ago      314MB
mcr.microsoft.com/dotnet/core/runtime   2.2                 d51bb4452469        2 days ago          314MB
```

Each command in the *Dockerfile* generated a layer and created an **IMAGE ID**. The final **IMAGE ID** (yours will be different) is **ddcc6646461b** and next you'll create a container based on this image.

## Create a container

Now that you have an image that contains your app, you can create a container. You can create a container in two ways. First, create a new container that is stopped.

```console
> docker create myimage
0e8f3c2ca32ce773712a5cca38750f41259a4e54e04bdf0946087e230ad7066c
```

> The `docker create` command from above will create a container based on the **myimage** image. The output of that command shows you the **CONTAINER ID** of the created container. To see a list of *all* containers, use the `docker ps -a` command:

Reviewer: Like with images, you should mention that the container ID will be different.

```console
> docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS        PORTS   NAMES
0e8f3c2ca32c        myimage             "dotnet app/myapp.dll"   4 seconds ago       Created               boring_matsumoto
```

### Manage the container

> Each container is assigned a random name that you can use to refer to that container instance. For example, the container that was created automatically chose the name **boring_matsumoto** and that name can be used to start the container. The following example uses the `docker start` command to start the container, and then uses the `docker ps` command to only show containers that are running:

Reviewer: I'd mention that you can name a container when you create (or run) it. I wouldn't change the flow of this tutorial, but I think that would anticipate reader questions.

0e8f3c2ca32c        myimage             "dotnet app/myapp.dll"   7 minutes ago       Up 8 seconds           boring_matsumoto
```

Similarly, the `docker stop` command will stop the container. The following example uses the `docker stop` command to stop the container, and then uses the `docker ps` command to show that no containers are running.

```console
> docker stop boring_matsumoto
boring_matsumoto

> docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS     PORTS   NAMES
```

### Connect to a container

After a container is running, you can connect to it to see the output. Use the `docker start` and `docker attach` commands to start the container and peek at the output stream. In this example, the <kbd>CTRL + C</kbd> command is used to detach from the running container. Reattach to the container to verify that it's still running and counting.

```console
> docker start boring_matsumoto
boring_matsumoto

> docker attach boring_matsumoto
Counter: 7
Counter: 8
Counter: 9
^C

> docker attach boring_matsumoto
Counter: 17
Counter: 18
Counter: 19
^C
```

### Delete a container

> The the purposes of this article you don't want containers just hanging around doing nothing. Delete the container you previously created. If the container is running, stop it.

Reviewer: ```suggestion
For the purposes of this article you don't want containers just hanging around doing nothing. Delete the container you previously created. If the container is running, stop it.
```

* [docker ps](https://docs.docker.com/engine/reference/commandline/ps/)
* [docker stop](https://docs.docker.com/engine/reference/commandline/stop/)
* [docker rm](https://docs.docker.com/engine/reference/commandline/rm/)
* [docker rmi](https://docs.docker.com/engine/reference/commandline/rmi/)
* [docker image](https://docs.docker.com/engine/reference/commandline/image/)

## Clean up resources

During this tutorial you created containers and images. If you want, delete these resources. Use the following commands to

01. List all containers

    ```console
    > docker ps -a
    ```

02. Stop containers that are running. The `CONTAINER_NAME` represents the name automatically assigned to the container.

    ```console
    > docker stop CONTAINER_NAME
    ```

03. Delete the container

    ```console
    > docker rm CONTAINER_NAME
    ```

Next, delete any images that you no longer want on your machine. Delete the image created by your *Dockerfile* and then delete the .NET Core image the *Dockerfile* was based on. You can use the **IMAGE ID** or the **REPOSITORY:TAG** formatted string.

```console
docker rmi myimage:latest

> docker rmi mcr.microsoft.com/dotnet/core/runtime:2.2

Reviewer: Consider the following suggestion (maybe as a NOTE):

> Image files can be large. Typically, you remove images you created while developing or testing an application. You usually keep the base image with the runtime if you plan on building other images based on that runtime.
